How to store forms data in Azure Storage (#2875)
umbraco/UmbracoDocs: Extending/FileSystemProviders/Azure-Blob-Storage/index-v7.md

versionFrom: 7.0.0
---

# Setup Your Site to use Azure Blob Storage for Media, Forms data and Image Processor Cache

> For Umbraco sites there are some scenarios when you may want, or need, to consider using Azure Blob Storage for your media and forms.  Particularly if your site contains large amounts of media.  Having your site’s media in Azure Storage can also help your deployments complete more quickly and has the potential to positively affect site performance as the Image Processor cache is moved to Azure Blob Storage.  It also allows you to serve your media from the Azure CDN. Moving your forms data to Azure Storage allows you to have multiple environments share the same forms data folder. This particularly useful when using Web App Slots in Azure App Service.

Reviewer: ```suggestion
For Umbraco sites there are some scenarios when you may want or need, to consider using Azure Blob Storage for your media and forms.  Particularly if your site contains large amounts of media.  Having your site’s media in Azure Storage can also help your deployments complete more quickly and have the potential to positively affect site performance as the Image Processor cache is moved to Azure Blob Storage.  It also allows you to serve your media from the Azure CDN. Moving your forms data to Azure Blob Storage allows you to have multiple environments share the same forms data folder. This is particularly useful when using Web App Slots in Azure App Service.
```

    </location>
  </configuration>
```
### Forms

> To clarify we are not storing form submissions into Azure storage but the JSON files used to map the layout of the forms such as the templates used and workflows. Forms submissions are stored in the database.

Reviewer: ```suggestion
We are not storing form submissions into Azure Blob Storage but the JSON files used to map the layout of the forms such as the templates used and workflows. Forms submissions are stored in the database.
```

> To configure Umbraco forms data its very similar to how media is setup above. You just need to provide another provider as a sibling to the media provider. In the snippet below I have added a provider with the alias of forms along side with the media provider.

Reviewer: ```suggestion
Configuring Umbraco Forms data is very similar to how media is setup above. You need to provide another provider as a sibling to the media provider. In the snippet below I have added a provider with the alias of "forms" alongside the media provider.
```

> ```xml

Reviewer: I am not entirely sure I understand this code snippet here.
There is ONLY a provider for the forms here, right?
Looks like the media provider is closed in the same line as it's started.

I might be wrong though 🤔
Perhaps you could help me understand exactly what you're doing here?

Author: To keep the snippet short I removed all the xml code inside the media provider to just show the bare bones of it along side the forms provider.

Reviewer: Alright! Makes sense 💪 
Perhaps that could be explained in [a note](https://our.umbraco.com/documentation/Contribute/Markdown-Conventions/#adding-notes-warnings-tips) before or after the snippet?
What do you think?

<?xml version="1.0"?>
<FileSystemProviders>
  <!-- Media -->
  <Provider alias="media" type="Our.Umbraco.FileSystemProviders.Azure.AzureBlobFileSystem, Our.Umbraco.FileSystemProviders.Azure"></Provider>
  <!-- Forms -->
	<Provider alias="forms" type="Our.Umbraco.FileSystemProviders.Azure.AzureBlobFileSystem, Our.Umbraco.FileSystemProviders.Azure">
    <Parameters>
    <add key="containerName" value="forms"/>
    <add key="rootUrl" value="http://[myAccountName].blob.core.windows.net/"/>
    <add key="connectionString" value="DefaultEndpointsProtocol=https;AccountName=[myAccountName];AccountKey=[myAccountKey]"/>
    <!--
        Optional configuration value determining the maximum number of days to cache items in the browser.
        Defaults to 365 days.
    -->
    <add key="maxDays" value="365"/>
    <!--
        When true this allows the VirtualPathProvider to use the default "media" route prefix regardless
        of the container name.
    -->
    <add key="useDefaultRoute" value="true"/>
    <!--
        When true blob containers will be private instead of public what means that you can't access the original blob file directly from its blob url.
    -->

>     <add key="usePrivateContainer" value="true"/>

Reviewer: So, setting this to `true` means that the folder can't be accessed? Is that correct?

Author: Yes forms data shouldn't be publicly accessible unlike media. By setting this to true we confirm that we are using private blob containers.

Any new media files you upload to the site, will automatically be added to the Blob Storage.

## Exisiting Forms files

> Any Umbraco forms data files will not automatically be added to blob storage. You will need to upload the **contents** of the Umbraco forms data folder to storage. This can be found in `~/App_Plugins/UmbracoForms/Data`. Again you need the **contents** of the data folder so the **forms**, **templates** and **workflows** folders should be the only folders in the container. Do not upload the whole data folder! 

Reviewer: ```suggestion
Any existing Umbraco Forms data files will not automatically be added to the blob storage. You will need to upload the **contents** of the Umbraco Forms data folder to the storage. This can be found in `~/App_Plugins/UmbracoForms/Data`. Again you need the **contents** of the Data folder so the **forms**, **templates** and **workflows** folders should be the only folders in the container. Do not upload the whole Data folder. 
```

Author: "the blob storage" and "the storage" makes sense but sounds wrong. I guess changing it to be this would be better: "automatically be added to Azure Blob storage" "Umbraco Forms data folder to Blob storage". 

I am not sure how to change your suggestion. "Add suggestion to batch" is greyed out for me

Reviewer: I see your point 🤔
I've tried to update the wording a bit..
Instead of saying "the Blob storage", I've left it out all together, so it now says: "You will need to upload the contents of the Umbraco Forms data folder manually."

I think that works better. What do you think?
